File: Android/dokit-gps-mock/src/main/java/com/didichuxing/doraemonkit/gps_mock/gpsmock/WifiHooker.java
package com.didichuxing.doraemonkit.gps_mock.gpsmock;

import android.content.Context;
import android.net.wifi.ScanResult;
import android.net.wifi.WifiManager;
import android.os.IBinder;

import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by wanglikun on 2019/4/2
 */
public class WifiHooker extends BaseServiceHooker {
    @Override
    public String serviceName() {
        return Context.WIFI_SERVICE;
    }

    @Override
    public String stubName() {
        return "android.net.wifi.IWifiManager$Stub";
    }

    @Override
    public Map<String, MethodHandler> registerMethodHandlers() {
        Map<String, MethodHandler> methodHandlers = new HashMap<>();
        methodHandlers.put("getScanResults", new GetScanResultsMethodHandler());
        methodHandlers.put("getConnectionInfo", new GetConnectionInfoMethodHandler());
        return methodHandlers;
    }

    @Override
    public void replaceBinderProxy(Context context, IBinder proxy) throws NoSuchFieldException, IllegalAccessException, ClassNotFoundException, NoSuchMethodException, InvocationTargetException {
        WifiManager wifiManager = (WifiManager) context.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        if (wifiManager == null) {
            return;
        }
        Class<?> wifiManagerClass = wifiManager.getClass();
        Field mServiceField = wifiManagerClass.getDeclaredField("mService");
        mServiceField.setAccessible(true);
        Class stub = Class.forName(stubName());
        Method asInterface = stub.getDeclaredMethod("asInterface", IBinder.class);
        mServiceField.set(wifiManager, asInterface.invoke(null, proxy));
        mServiceField.setAccessible(false);
    }

    static class GetScanResultsMethodHandler extends MethodHandler {

        @Override
        public Object onInvoke(Object originService, Method method, Object[] args) throws InvocationTargetException, IllegalAccessException {
            if (!GpsMockManager.getInstance().isMocking()) {
                return method.invoke(originService, args);
            }
            return new ArrayList<ScanResult>();
        }


    }

    static class GetConnectionInfoMethodHandler extends MethodHandler {

        @Override
        public Object onInvoke(Object originObject, Method method, Object[] args) throws InvocationTargetException, IllegalAccessException {
            if (!GpsMockManager.getInstance().isMocking()) {
                return method.invoke(originObject, args);
            }
            try {
                return Class.forName("android.net.wifi.WifiInfo").newInstance();
            } catch (InstantiationException e) {
                e.printStackTrace();
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            }
            return method.invoke(originObject, args);
        }
    }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/hook/WifiHooker.kt
package com.didichuxing.doraemonkit.hook

import android.content.Context
import android.net.wifi.ScanResult
import android.net.wifi.WifiManager
import android.os.IBinder
import com.didichuxing.doraemonkit.kit.gpsmock.GpsMockManager
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.*

/**
 * Created by wanglikun on 2019/4/2
 */
class WifiHooker : BaseServiceHooker() {
    override val serviceName: String
        get() = Context.WIFI_SERVICE

    override val stubName: String
        get() = "android.net.wifi.IWifiManager\$Stub"

    override val methodHandlers: MutableMap<String, MethodHandler?>
        get() {
            val methodHandlers: MutableMap<String, MethodHandler?> = mutableMapOf()
            methodHandlers["getScanResults"] = GetScanResultsMethodHandler()
            methodHandlers["getConnectionInfo"] = GetConnectionInfoMethodHandler()
            return methodHandlers
        }

    @Throws(NoSuchFieldException::class, IllegalAccessException::class, ClassNotFoundException::class, NoSuchMethodException::class, InvocationTargetException::class)
    override fun replaceBinder(context: Context?, proxy: IBinder?) {
        val wifiManager = context!!.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
                ?: return
        val wifiManagerClass: Class<*> = wifiManager.javaClass
        val mServiceField = wifiManagerClass.getDeclaredField("mService")
        mServiceField.isAccessible = true
        val stub = Class.forName(stubName)
        val asInterface = stub.getDeclaredMethod(BaseServiceHooker.METHOD_ASINTERFACE, IBinder::class.java)
        mServiceField[wifiManager] = asInterface.invoke(null, proxy)
        mServiceField.isAccessible = false
    }

    internal class GetScanResultsMethodHandler : MethodHandler {
        @Throws(InvocationTargetException::class, IllegalAccessException::class)
        override fun onInvoke(originService: Any?, proxy: Any?, method: Method?, args: Array<Any>?): Any? {
            return if (!GpsMockManager.instance.isMocking) {
                method!!.invoke(originService, args)
            } else ArrayList<ScanResult>()
        }
    }

    internal class GetConnectionInfoMethodHandler : MethodHandler {
        @Throws(InvocationTargetException::class, IllegalAccessException::class)
        override fun onInvoke(originObject: Any?, proxyObject: Any?, method: Method?, args: Array<Any>?): Any? {
            if (!GpsMockManager.instance.isMocking) {
                return method!!.invoke(originObject, args)
            }
            try {
                return Class.forName("android.net.wifi.WifiInfo").newInstance()
            } catch (e: InstantiationException) {
                e.printStackTrace()
            } catch (e: ClassNotFoundException) {
                e.printStackTrace()
            }
            return method!!.invoke(originObject, args)
        }
    }
}